experimental/rule_inference/utils/rule_utils.py

```diff
@@ -1,44 +1,140 @@
 import json
-from typing import Dict, List
+from typing import Dict, List, Set
+import attr
 
-from comby import Comby
 from polyglot_piranha import Filter, OutgoingEdges, Rule, RuleGraph
 
 
+@attr.s(eq=False)
+class RawFilter:
+    id = attr.ib(type=str, default="0")
+    enclosing_node = attr.ib(type=str, default=None)
+    not_enclosing_node = attr.ib(type=str, default=None)
+    not_contains = attr.ib(type=List[str], default=[])
+    contains = attr.ib(type=str, default=None)
+    at_least = attr.ib(type=int, default=1)
+    at_most = attr.ib(type=int, default=4294967295)  # u32::MAX
+    child_count = attr.ib(type=int, default=4294967295)  # u32::MAX
+    sibling_count = attr.ib(type=int, default=4294967295)  # u32::MAX
+
+    def to_toml(self):
+        str_reps = []
+        if self.enclosing_node:
+            str_reps.append(f'enclosing_node = """{self.enclosing_node}"""')
+        if self.not_enclosing_node:
+            str_reps.append(f'not_enclosing_node = """{self.not_enclosing_node}"""')
+        if self.not_contains:
+            str_reps.append(f"not_contains = {json.dumps(self.not_contains)}")
+        if self.contains:
+            str_reps.append(f'contains = """{self.contains}"""')
+        if self.at_least != 1:
+            str_reps.append(f"at_least = {self.at_least}")
+        if self.at_most != 4294967295:
+            str_reps.append(f"at_most = {self.at_most}")
+        if self.child_count != 4294967295:
+            str_reps.append(f"child_count = {self.child_count}")
+        if self.sibling_count != 4294967295:
+            str_reps.append(f"sibling_count = {self.sibling_count}")
+
+        return "\n".join([f"[[rules.filters]]"] + str_reps)
+
+    def to_filter(self):
+        return Filter(
+            enclosing_node=self.enclosing_node,
+            not_enclosing_node=self.not_enclosing_node,
+            not_contains=self.not_contains,
+            contains=self.contains,
+            at_least=self.at_least,
+            at_most=self.at_most,
+            # child_count=self.child_count,
```

Comment:
why is this commented out?

Reply:
My version of piranha didn;t support this